### PR TITLE
Factor ComputeSpacetimeQuantities and ComputeGhQuantities

### PR DESCRIPTION
**Pull request by geoffrey4444**
## Proposed changes

Addresses #2212 . Factors PointwiseFunctions/GeneralRelativity/ComputeSpacetimeQuantities.?pp into multiple files, and factors PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.?pp into multiple files in PointwiseFunctions/GeneralRelativity/GeneralizedHarmonic/.

UPDATE: updated to factor both ComputeSpacetimeQuantities and ComputeGhQuantities
### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [x] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

**Comment by geoffrey4444** [comment]
Updated so clang-tidy is happy

**Comment by geoffrey4444** [comment]
Updated so clang-tidy is happy

File: src/PointwiseFunctions/GeneralRelativity/SpacetimeMetric.cpp
Comment (moxcodes) [inline src/PointwiseFunctions/GeneralRelativity/SpacetimeMetric.cpp:20]: There are a number of minor tweaks I'd suggest to modernize some of these functions, like using `destructive_resize` and initializing with the size in the return-by-value version to avoid forcing the control flow into the `UNLIKELY` branch, but I think it's probably best to leave it be to ensure that these changes are purely moving code so that  the review is easier.
Perhaps I should just make an issue to revisit updating these functions with those code tweaks we've been using?

File: src/PointwiseFunctions/GeneralRelativity/DerivativesOfSpacetimeMetric.hpp
Comment (moxcodes) [inline src/PointwiseFunctions/GeneralRelativity/DerivativesOfSpacetimeMetric.hpp:43]: (in a separate commit) maybe consider `spacetime_derivative_of_spacetime_metric`, as the plural could be misinterpreted to mean that the function has multiple returns by pointer, or returns multiple things in a `std::pair`.

File: src/PointwiseFunctions/GeneralRelativity/GeneralizedHarmonic/DerivSpatialMetric.hpp
Comment (moxcodes) [inline src/PointwiseFunctions/GeneralRelativity/GeneralizedHarmonic/DerivSpatialMetric.hpp:50]: in most other cases, we've been careful to specify `spatial_derivative`, `time_derivative` or `spacetime_derivative`, perhaps consider changing the name (again, in a separate commit to keep this a pure-move)?

File: src/PointwiseFunctions/GeneralRelativity/GeneralizedHarmonic/SpacetimeDerivOfNormOfShift.hpp
Comment (moxcodes) [inline src/PointwiseFunctions/GeneralRelativity/GeneralizedHarmonic/SpacetimeDerivOfNormOfShift.hpp:19]: unless you've run this through IWYU after the changes, let's just get rid of the IWYU directives.

File: src/PointwiseFunctions/GeneralRelativity/GeneralizedHarmonic/SpatialDerivOfLapse.hpp
Comment (moxcodes) [inline src/PointwiseFunctions/GeneralRelativity/GeneralizedHarmonic/SpatialDerivOfLapse.hpp:82]: maybe rename the compute tag to be more specific

File: src/PointwiseFunctions/GeneralRelativity/GeneralizedHarmonic/CMakeLists.txt
Comment (moxcodes) [inline src/PointwiseFunctions/GeneralRelativity/GeneralizedHarmonic/CMakeLists.txt:13]: alphabet - `DerivOf` should get sourted to before `Derivative` because caps < lower case

File: src/PointwiseFunctions/GeneralRelativity/CMakeLists.txt
Comment (moxcodes) [inline src/PointwiseFunctions/GeneralRelativity/CMakeLists.txt:23]: for my understanding, why are the headers here not declared like the `GeneralizedHarmonic` ones?

**Review by moxcodes — COMMENTED** [review]
I've verified that the factoring commits:
- are one-to-one and onto (introduce no new functions and lose no functions, and the moved functions are identical)
- all of the moved functions' documentation renders
So largely I'm fine with these changes, I just have a couple of small suggestions and questions.

File: src/PointwiseFunctions/GeneralRelativity/SpacetimeMetric.cpp
Comment (geoffrey4444) [reply src/PointwiseFunctions/GeneralRelativity/SpacetimeMetric.cpp:20]: Please open an issue and assign to me. I’d like to do anything non-moving as a follow up PR

File: src/PointwiseFunctions/GeneralRelativity/GeneralizedHarmonic/SpatialDerivOfLapse.hpp
Comment (geoffrey4444) [reply src/PointwiseFunctions/GeneralRelativity/GeneralizedHarmonic/SpatialDerivOfLapse.hpp:82]: In a follow up commit?

Comment (moxcodes) [reply src/PointwiseFunctions/GeneralRelativity/GeneralizedHarmonic/SpatialDerivOfLapse.hpp:82]: yes, please

File: src/PointwiseFunctions/GeneralRelativity/SpacetimeMetric.cpp
Comment (moxcodes) [reply src/PointwiseFunctions/GeneralRelativity/SpacetimeMetric.cpp:20]: created issue #2350

File: src/PointwiseFunctions/GeneralRelativity/CMakeLists.txt
Comment (nilsdeppe) [reply src/PointwiseFunctions/GeneralRelativity/CMakeLists.txt:23]: They should be. We are in the process of switching over to declaring the headers (@fmahebert is doing a lot of the work here, which I am extremely appreciative of!!!). In the future we will need to keep track of all source and header files. It would be nice to add them already, but see if it conflicts with any modernization PRs of CMake that are open

File: src/PointwiseFunctions/GeneralRelativity/GeneralizedHarmonic/SpacetimeDerivOfNormOfShift.hpp
Comment (nilsdeppe) [reply src/PointwiseFunctions/GeneralRelativity/GeneralizedHarmonic/SpacetimeDerivOfNormOfShift.hpp:19]: Agreed

**Review by nilsdeppe — DISMISSED** [review]
I agree with @moxcodes suggestions, but I am also very sympathetic to the number of conflicts and work that is quickly created having such a large refactor PR open. My advice would be to merge this, and then have @geoffrey4444 address the comments not directly related to the refactoring in one or more followup PRs. I'm very appreciative of more people helping out with code maintenance and refactoring things like this so I don't want to make their life any more difficult than it absolutely needs to be.

**Comment by geoffrey4444** [comment]
> I agree with @moxcodes suggestions, but I am also very sympathetic to the number of conflicts and work that is quickly created having such a large refactor PR open. My advice would be to merge this, and then have @geoffrey4444 address the comments not directly related to the refactoring in one or more followup PRs. I'm very appreciative of more people helping out with code maintenance and refactoring things like this so I don't want to make their life any more difficult than it absolutely needs to be.

I appreciate it, @nilsdeppe ! If this gets merged I’ll do the fix ups ASAP in a follow up commit

**Review by moxcodes — DISMISSED** [review]
@geoffrey4444 @nilsdeppe I'm happy with the suggestion to just proceed with merging this to minimize the labor of working on these files, so we can just consider my comments as suggestions for the future tune-up of these functions.
It does look like there are already conflicts, so I think this will require one rebase before it can be merged.

File: src/PointwiseFunctions/GeneralRelativity/CMakeLists.txt
Comment (fmahebert) [reply src/PointwiseFunctions/GeneralRelativity/CMakeLists.txt:23]: I have not yet touched any of the CMakeLists.txt in `PointwiseFunctions`, partly because I was waiting for this PR to get merged in first. For me it makes little difference to update one additional CMakeLists.txt file, so I would be happy for us to merge @geoffrey4444's PR as-is and I will modernize the CMake afterwards. This should in particular help with avoiding conflicts from keeping this large refactoring PR open any longer than necessary.

But if @geoffrey4444 wants a try at updating a CMakeLists.txt, that's great too 🙂 

@moxcodes which GeneralizedHarmonic headers are you referring to?

**Comment by nilsdeppe** [comment]
Great! @geoffrey4444 please ping me whenever you've had a chance to rebase so we can merge this :)

**Comment by geoffrey4444** [comment]
> Great! @geoffrey4444 please ping me whenever you've had a chance to rebase so we can merge this :)

I will! Thanks!

**Comment by geoffrey4444** [comment]
Rebased @nilsdeppe

**Comment by geoffrey4444** [comment]
Just squashed in another clang-tidy fix that got triggered after rebasing

**Review by moxcodes — APPROVED** [review]

**Comment by geoffrey4444** [comment]
Test failure is a random hydro test failure...can I losses its tolerance in a follow up PR? I’m happy to have an issue assigned to me

**Comment by nilsdeppe** [comment]
Yea, just ignore that. Feel free to just make an issue and I'll see who is the right person to assign it to

**Review by nilsdeppe — APPROVED** [review]

**Pull request merged**
